pkgs/build-support/node/fetch-yarn-deps/yarn-berry-supported-archs.json

```diff
@@ -0,0 +1,16 @@
+{
```

Comment:
This file means that we can never expand our platform support without mass FOD breaks across the tree? That seems very bad. There are already platforms you can boot NixOS on that would be ruled out by this file. At the very least we should try to list every value already used in npm here so that it’s relatively safe to expand in future?

Reply:
> This file means that we can never expand our platform support without mass FOD breaks across the tree? 

yes

>That seems very bad. There are already platforms you can boot NixOS on that would be ruled out by this file. At the very least we should try to list every value already used in npm here so that it’s relatively safe to expand in future?

We did add all the relevant platforms, cpu's and libc's (see https://yarnpkg.com/configuration/yarnrc#supportedArchitectures) that are available. The only one missing is `win32` and I don't think we need to add this

Reply:
If it’s everything Yarn supports then I guess it’s not so bad – assuming we will be able to add additional platforms if Yarn does support more in future without invalidating every FOD?

I do think we should include Windows. You can use Nixpkgs to cross‐build a lot of stuff for Windows today, and people are working on native bring‐up. What’s the harm?

Reply:
ok, lets add `win32`.
If yarn adds support for further platforms, we can add those, but we will need to re-validate all of the hashes involved. Hopefully this will be very rarely the case..

Reply:
I thought the idea is that packages can’t include any platforms other than the ones Yarn supports, so no existing FOD should change as long as we keep up with Yarn? Or can packages on npm reference other platforms and Yarn just can’t support fetching them?

Reply:
done. 
On a sidenode: I did neither have to change the test hashes, nor the hash for the (seriously complex and long) yarn pgadmin offlineCache. (and yes, I obviously invalidated them, re-downloaded the packages and compared the hashes)
So maybe this won't be such a big deal after all

Reply:
> I thought the idea is that packages can’t include any platforms other than the ones Yarn supports, so no existing FOD should change as long as we keep up with Yarn? 

Yes, I think your right. My guess is that adding another platform or compiler will a huge undertaking anyway, so maybe they will publish yet another big version upgrade, which would make it easier for us (e.g. `yarn-berry_5`)

Reply:
Unfortunately, I think I was correct here:

> Or can packages on npm reference other platforms and Yarn just can’t support fetching them?

The npm docs imply that the fields are quite freeform; e.g. the [`cpu` example](https://docs.npmjs.com/cli/v10/configuring-npm/package-json#cpu) uses `mips` as an example. That means that this is just a limitation of Yarn’s configuration format rather than a constraint on the actual packages in existence, and there are packages that could change in the future if we add new platforms like RISC‐V.

The correct solution here would be to fetch everything independent of platform – theoretically Yarn could add `all` values here to opt in to fetching everything, but writing our own deterministic fetcher would make it easy and also obviate the other reproducibility concerns. If neither of those is an option, perhaps we can find a way to check that nothing in the dependency tree references any platform we haven’t included in this file, and otherwise bail out? If we can’t do even that, then it’d be good if someone could write a script to scan the npm repository and collect every value that is used in practice so that we can include them all from the start and reduce the scope of potential changes.

Because we’re adding fundamental packaging infrastructure here that is part of our public interface and will surely be relied upon by out‐of‐tree packages, and changing FOD hashes are very difficult to spot and debug, I don’t think we’ll have the option of *just breaking it* in the future, unless we can show that the backwards incompatibility would be entirely theoretical in the existing public corpus.

Reply:
> Unfortunately, I think I was correct here:
> 
> > Or can packages on npm reference other platforms and Yarn just can’t support fetching them?
> 
> The npm docs imply that the fields are quite freeform; e.g. the [`cpu` example](https://docs.npmjs.com/cli/v10/configuring-npm/package-json#cpu) uses `mips` as an example. That means that this is just a limitation of Yarn’s configuration format rather than a constraint on the actual packages in existence, and there are packages that could change in the future if we add new platforms like RISC‐V.
> 

Actually, it is not a limitation of the configuration, as it can be adapted to include everything node supports. https://yarnpkg.com/configuration/yarnrc#supportedArchitectures says:

>List of CPU architectures to cover.
See https://nodejs.org/docs/latest/api/process.html#processarch for the architectures supported by Node.js

which states:
>The operating system CPU architecture for which the Node.js binary was compiled. Possible values are: 'arm', 'arm64', 'ia32', 'loong64', 'mips', 'mipsel', 'ppc', 'ppc64', 'riscv64', 's390', 's390x', and 'x64'.

So we can include all of those for our `supported-archs`. Adding this, btw, didn't change the cache for pgadmin, either.
I looked into how `yarn-berry` does its comparisons (since these values are for the local `.yarnrc.toml` config file) and it compares it to upstreams `package.json` file. This file has the same fields (https://yarnpkg.com/configuration/manifest#cpu) and, more importantly, _only_ the `libc` and `os` fields we already have.

> The correct solution here would be to fetch everything independent of platform – theoretically Yarn could add `all` values here to opt in to fetching everything, but writing our own deterministic fetcher would make it easy and also obviate the other reproducibility concerns. 

Yes, writing our own fetcher could be a solution. This will be quite a task to parse the `yarn.lock` file and generate the cache. I really feel uncomftable reinventing the wheel, though. We do have a package manager, which already parses this file and also controlls the specs about our architectures. I don't want to write yet another solution of `yarn2nix`

>If neither of those is an option, perhaps we can find a way to check that nothing in the dependency tree references any platform we haven’t included in this file, and otherwise bail out? 

There seems to be an option (`yarn info --all --manifest --json`) which prints all the manifests, but _not one_ had a `cpu`, `libc` or `os` field. There is another option (`yarn npm info _packageName_ --json`) which prints the upstream manifest and from the quick peek I took, didn't include any of those, either.

>If we can’t do even that, then it’d be good if someone could write a script to scan the npm repository and collect every value that is used in practice so that we can include them all from the start and reduce the scope of potential changes.
> Because we’re adding fundamental packaging infrastructure here that is part of our public interface and will surely be relied upon by out‐of‐tree packages, and changing FOD hashes are very difficult to spot and debug, I don’t think we’ll have the option of _just breaking it_ in the future, unless we can show that the backwards incompatibility would be entirely theoretical in the existing public corpus.

Totally agree. This is why we added versions here. Not only because upstream has two supported versions (3 and 4), but also because their `yarn.lock` files differ. 
If there is a situation where either the file format changes, or the supported architectures, I would suggest adding another version (e.g. `yarn-berry_5` or `yarn-berry_39`). This will keep all the hashes intakt, be reproducible and still be able to adapt to future changes.

Reply:
> The npm docs imply that the fields are quite freeform; e.g. the [`cpu` example](https://docs.npmjs.com/cli/v10/configuring-npm/package-json#cpu) uses `mips` as an example. That means that this is just a limitation of Yarn’s configuration format rather than a constraint on the actual packages in existence, and there are packages that could change in the future if we add new platforms like RISC‐V.

The documentation says these values are based on Node `process.{arch,cpu}` which is a known [documented](https://nodejs.org/api/process.html#processarch) set of values. I wouldn't exactly call it freeform.

I don't really have a better idea on how to handle adding new platform other than invalidating the FOD (by changing `pname`) and fixing up all the builds or introducing a new builder version.

I wonder how `pnpm` managed to avoid this issue. I don't see it handling `supportedArchitectures` at all.